python/src/dataset.rs

```diff
@@ -589,6 +592,36 @@ impl Dataset {
             scanner.fragment_readahead(fragment_readahead);
         }
 
+        if let Some(stats_handler) = stats_handler {
+            if stats_handler.downcast::<PyString>().is_ok() {
+                let stats_handler = stats_handler.extract::<String>()?;
+                scanner.stats_handler(
+                    stats_handler
+                        .parse::<ScanStatisticsHandler>()
+                        .infer_error()?,
+                );
+            } else if stats_handler.is_callable() {
+                let stats_handler = stats_handler.unbind();
+                let stats_handler = ScanStatisticsHandler::Custom(Arc::new(move |stats| {
+                    let wrapped_stats = LanceScanStats::new(stats);
+                    Python::with_gil(|py| {
+                        let args = PyTuple::new_bound(py, vec![wrapped_stats.into_py(py)]);
+                        stats_handler.call1(py, args)
+                    })
+                    .map_err(|err| lance_core::Error::Wrapped {
+                        error: err.into(),
+                        location: location!(),
+                    })?;
```

Comment:
I might recommend using `format_python_error()` error here. The nice thing it does is puts the traceback in the error message, which helps Python users a lot in identifying the source of an error.